content/blog/2019-09-25-c-implementation.md

```diff
@@ -0,0 +1,189 @@
+program posted in the class slack channel, on which the reference interpreter
+gave incorrect results. 
+
+### Performance
+I wanted to have a way to get reliable performance numbers to square off
+againts any other student's implementation of bril. First, I had the main
+evaluation loop return a `uint64_t` number of nanoseconds of elapsed time.
+I used the POSIX provided `timespec` structure to record the time tracked by
+`CLOCK_PROCESS_CPUTIME_ID`, the nanosecond resolution process time clock. This
+tracks CPU ticks spent on the program process itself, irrespective of other
+scheduled processes. The code is reproduced below:
+```c
+  JSON_Array *instrs = get_main_func_instrs(fn);
+  struct timespec tick, tock;
+  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &tick);
+  mem = table_init();
+  labels = table_init();
+  make_dispatch();
+
+  num_instrs = json_array_get_count(instrs);
+  
+  for (size_t i = 0; i < num_instrs; ++i) {
+    curr_instr = json_array_get_object(instrs, i);
+    const char *label = json_object_get_string(curr_instr, "label");
+    if (label) {
+      table_put(labels, label, i);
+    }
+  }
+
+  size_t op_ind = -1;
+  char const *op = 0;
+  bool not_found = false;
+  for (ip = 0; ip < num_instrs; ++ip) {
+    curr_instr = json_array_get_object(instrs, ip);
+    op = jogs(curr_instr, "op");
+    if (!op) continue; // label
+    op_ind = (size_t) table_get(disp, op, &not_found);
+    if (not_found) {
+      quit("Operation not defined", op);
+    } else {
+      (op_funcs[op_ind]());
+    }
+  }
+
+  cleanup();
+  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &tock);
+  return (1000000000 * (tock.tv_sec - tick.tv_sec) + tock.tv_nsec - tick.tv_nsec);
+```
+
+Notice the following:
+* I start timing on line 2, after JSON parsing. I did not want that to be
+  included
+* I stop timing AFTER cleaning up all data structures used for interpretation
+* The last line returns the time elapsed in nanoseconds
+
+Then, in `src/main.c`, I have a constant named `NUM_RUNS`, and if `cril` is
+called with `--benchmark`, it will run each program under `tests` `NUM_RUNS`
+times, calculate a mean and standard deviation per program, and output the
+results. I've included some sample output for `NUM_RUNS=10000"
+
+```
+NUM RUNS: 10000
+test/interp/
+--------------------------------------
+test/interp/jmp.json
+MEAN: 0.004536
+STDV: 0.003142
+
+test/interp/div.json
+MEAN: 0.003799
+STDV: 0.001767
+
+test/interp/ret.json
+MEAN: 0.003404
+STDV: 0.001907
+
+test/interp/nop.json
+MEAN: 0.003787
+STDV: 0.002202
+
+test/interp/br.json
+MEAN: 0.004586
+STDV: 0.002476
+
+test/interp/fib.json
+MEAN: 0.108300
+STDV: 0.029558
+
+test/parse/
+--------------------------------------
+test/parse/print.json
+MEAN: 0.003200
+STDV: 0.002105
+
+test/parse/comment.json
+MEAN: 0.002641
+STDV: 0.001431
+
+test/parse/add.json
+MEAN: 0.003943
+STDV: 0.002204
+
+test/print/
+--------------------------------------
+test/print/add.json
+MEAN: 0.004013
+STDV: 0.002048
+```
```

Comment:
Comparing the performance of programs as simple as `jmp` and `add` probably isn't that meaningful. To get a useful point of comparison, you'll want to run programs that at least include a loop, IMO. So maybe focus on `fib` here and, if you're feeling so inclined, run it for a few different inputs?